### PR TITLE
Adding settings page

### PR DESCRIPTION
**Pull request by hayesjordan**

### CODE REVIEW

File: android/PhysicalWeb/app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -19,12 +19,17 @@
...
     <string name="title_about">About</string>
+    <string name="title_settings">Settings</string>
     <string name="empty_nearby_beacons_list_text">Searching for beacons</string>
     <string name="about_app_title">Physical Web</string>
     <string name="url_getting_started">        https://google.github.io/physical-web/mobile/android/getting-started.html</string>
     <string name="config_no_beacons_found">No beacons found</string>
     <string name="notification_view_more_beacons">VIEW MORE</string>
     <string name="config_url_read_error">Error Reading URL</string>
+    <string name="settings_pws_endpoint_selector">PWS Endpoint</string>
+    <string name="settings_prefs_key">org.physical_web.physicalweb.SETTINGS_PREFS</string>
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/res/values/strings.xml]: This can just be a constant in the file where it's used

File: android/PhysicalWeb/app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -19,12 +19,17 @@
...
     <string name="title_nearby_beacons">Nearby Beacons</string>
     <string name="title_about">About</string>
+    <string name="title_settings">Settings</string>
     <string name="empty_nearby_beacons_list_text">Searching for beacons</string>
     <string name="about_app_title">Physical Web</string>
     <string name="url_getting_started">        https://google.github.io/physical-web/mobile/android/getting-started.html</string>
     <string name="config_no_beacons_found">No beacons found</string>
     <string name="notification_view_more_beacons">VIEW MORE</string>
     <string name="config_url_read_error">Error Reading URL</string>
+    <string name="settings_pws_endpoint_selector">PWS Endpoint</string>
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/res/values/strings.xml]: Just make this "Physical Web Service"

File: android/PhysicalWeb/app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -19,12 +19,17 @@
...
+    <string name="settings_prefs_key">org.physical_web.physicalweb.SETTINGS_PREFS</string>
+    <string name="pws_endpoint_setting_display">PWS Endpoint</string>
+    <string name="pws_endpoint_setting_key">PWS Endpoint</string>
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/res/values/strings.xml]: This also can be a constant

File: android/PhysicalWeb/app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -19,12 +19,17 @@
...
+    <string name="title_settings">Settings</string>
     <string name="empty_nearby_beacons_list_text">Searching for beacons</string>
     <string name="about_app_title">Physical Web</string>
     <string name="url_getting_started">        https://google.github.io/physical-web/mobile/android/getting-started.html</string>
     <string name="config_no_beacons_found">No beacons found</string>
     <string name="notification_view_more_beacons">VIEW MORE</string>
     <string name="config_url_read_error">Error Reading URL</string>
+    <string name="settings_pws_endpoint_selector">PWS Endpoint</string>
+    <string name="settings_prefs_key">org.physical_web.physicalweb.SETTINGS_PREFS</string>
+    <string name="pws_endpoint_setting_display">PWS Endpoint</string>
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/res/values/strings.xml]: "Physical Web Service" is probably best here

File: android/PhysicalWeb/app/src/main/res/values/strings.xml (hunk 3 above)
Comment (cco3) [inline android/PhysicalWeb/app/src/main/res/values/strings.xml]: I'd prefer this were pws_endpoint (if this is a shared preference key)

File: android/PhysicalWeb/app/src/main/res/menu/main.xml [hunk 5]
```diff
@@ -13,4 +13,9 @@
           android:orderInCategory="100"
           android:id="@+id/action_about"/>
 
+    <item android:title="@string/action_settings"
+          app:showAsAction="never"
+          android:orderInCategory="100"
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/res/menu/main.xml]: Could you give it a non-duplicate order?  it should probably be less than the about (and then it should also come before the about in this xml file)

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  @Override
+  public void onResume() {
+    super.onResume();
+    getActivity().getActionBar().setTitle(R.string.title_settings);
+  }
+
+  @Override
+  public void onPause() {
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java]: You don't need to provide this

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+ */
+
+public class SettingsFragment extends PreferenceFragment{
+
+  private static final String TAG = "SettingsFragment";
+
+  public SettingsFragment() {
+  }
+
+  public static SettingsFragment newInstance() {
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java]: I understand that the beacon config fragment does this.  Do you know why?  Seems really weird...what is the normal practice?  Is it perhaps to return a singleton?

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    setHasOptionsMenu(true);
+    getActivity().getActionBar().setDisplayHomeAsUpEnabled(true);
+    // Load the preferences from an XML resource
+    addPreferencesFromResource(R.xml.settings);
+    ListPreference listPreference = (ListPreference) findPreference(
+      getString(R.string.pws_endpoint_setting_key));
+    CharSequence[] endpointNames;
+    CharSequence[] endpoints;
+    if (Utils.isGoogleApiKeyAvailable(getActivity())){
+      endpointNames = new CharSequence[]{"Google's PWS", "Prod PWS", "Dev PWS"};
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java]: These should be string resources

File: android/PhysicalWeb/app/src/main/res/menu/main.xml (hunk 5 above)
Comment (hayesjordan) [inline android/PhysicalWeb/app/src/main/res/menu/main.xml]: Should it come before or after the config?

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    super.onCreate(savedInstanceState);
+    setHasOptionsMenu(true);
+    getActivity().getActionBar().setDisplayHomeAsUpEnabled(true);
+    // Load the preferences from an XML resource
+    addPreferencesFromResource(R.xml.settings);
+    ListPreference listPreference = (ListPreference) findPreference(
+      getString(R.string.pws_endpoint_setting_key));
+    CharSequence[] endpointNames;
+    CharSequence[] endpoints;
+    if (Utils.isGoogleApiKeyAvailable(getActivity())){
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java]: Wouldn't it make sense to populate a list with the guaranteed two, and then just add the google pws if available?

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/Utils.java [hunk 10]
```diff
@@ -65,28 +67,48 @@ private static String getApiKey(Context context){
...
-      pwsEndpoint = GOOGLE_ENDPOINT;
-      apiVersion = GOOGLE_ENDPOINT_VERSION;
-    }
+    String pwsEndpoint = getCurrentPwsEndpointUrl(context);
+    int apiVersion = getCurrentPwsEndpointVersion(context);
     physicalWebCollection.setPwsEndpoint(pwsEndpoint, apiVersion, apiKey);
   }
 
+  public static boolean isGoogleApiKeyAvailable(Context context){
+    return context.getResources().getIdentifier("google_api_key", "string",
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/Utils.java]: Can you make a private `getGoogleApiKeyResourceId` that this and `getApiKey` uses?

File: android/PhysicalWeb/app/src/main/res/menu/main.xml (hunk 5 above)
Comment (cco3) [inline android/PhysicalWeb/app/src/main/res/menu/main.xml]: after

File: android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java (hunk 7 above)
Comment (hayesjordan) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java]: I also thought it was strange. It seems to have no real value. I just did it this way to be consistent. I think what I should do is remove it and call the constructor explicitly, then make another pull request to make the other fragments consistent. Thoughts?

Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java]: I don't mind if you change the other one in this change or later, but you could do a quick survey of other fragments in examples to see if they usually supply singletons or have any special instantiation methods?

Comment (hayesjordan) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java]: It seems `newInstance` was used for flexibility in the future. As seen in the link below Fragments can't have constructors with arguments.  What are your thoughts on keeping or scraping, if we keep I would add a comment for why it is there.

https://guides.codepath.com/android/Creating-and-Using-Fragments#fragment-with-arguments

Comment (cco3) [inline android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/SettingsFragment.java]: Just scrap it and the one in config fragment

**Comment by hayesjordan** [comment]
Updated per comments.

File: android/PhysicalWeb/app/src/main/res/values/strings.xml [hunk 11]
```diff
@@ -19,12 +19,18 @@
...
+    <string name="title_settings">Settings</string>
     <string name="empty_nearby_beacons_list_text">Searching for beacons</string>
     <string name="about_app_title">Physical Web</string>
     <string name="url_getting_started">        https://google.github.io/physical-web/mobile/android/getting-started.html</string>
     <string name="config_no_beacons_found">No beacons found</string>
     <string name="notification_view_more_beacons">VIEW MORE</string>
     <string name="config_url_read_error">Error Reading URL</string>
+    <string name="pws_endpoint_setting_display">Physical Web Service</string>
+    <string name="pws_endpoint_setting_key">pws_endpoint</string>
+    <string name="google_pws">Google\'s PWS</string>
```
Comment (cco3) [inline android/PhysicalWeb/app/src/main/res/values/strings.xml]: I'm thinking we should use these strings:
* Google Physical Web Service
* Production Open Source Physical Web Service
* Development Open Source Physical Web Service

Start with that, and maybe @scottjenson can comment.

**Comment by kevinahuber** [comment]
What are the differences between Google's PWS and Open Source PWS? This is a great add.

**Comment by scottjenson** [comment]
The open source PWS is mostly about caching the meta data for each page (Title, description, favicon) The Google version adds more sophisticated filtering, spam/phishing detection, and enforcement of HTTPS on the final destination page.

We hope others follow Opera's lead, fork our open source PWS and start implementing their own version. This is an ecosystem that is intended to work with many different PWS's in place.

**Comment by kevinahuber** [comment]
Is Google's PWS something that is open to the public to use? I am finding inconsistent behavior among which URL's work and which don't, and it would be very useful to have Google's PWS exposed to use and test. 

This is getting slightly off of the PR- would you prefer that I file an issue?

**Comment by scottjenson** [comment]
If my answer isn't enough, then yes, please open a separate issue. We are working on a testing front end to the PWS so people can test their URLs and get a result error if it fails (e.g. doesn't support HTTPS, robots.txt forbids us, etc) so you can test your URLs a bit easier. Coming in a few weeks.

**Comment by hayesjordan** [comment]
Updated

**Comment by cco3** [comment]
You are going to need to rebase this.  Talk to me if you need help with that.

**Comment by cco3** [comment]
LGTM, we can merge after travis checks it.

**Comment by cco3** [comment]
Sorry, one more comment, could you update the commit message to be in imperative mood?  I.e. "Add a settings page"

**Comment by hayesjordan** [comment]
Done

**Comment by cco3** [comment]
Great job!
